packages/core-data/src/entities.ts

```diff
@@ -12,157 +12,296 @@ import { __ } from '@wordpress/i18n';
 		name: 'globalStyles',
 		kind: 'root',
 		baseURL: '/wp/v2/global-styles',
 		baseURLParams: { context: 'edit' },
 		plural: 'globalStylesVariations', // should be different than name
 		getTitle: ( record ) => record?.title?.rendered || record?.title,
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Themes' ),
 		name: 'theme',
 		kind: 'root',
 		baseURL: '/wp/v2/themes',
 		baseURLParams: { context: 'edit' },
 		key: 'stylesheet',
-	},
-	{
+	} as const ),
+	defineEntity( {
 		label: __( 'Plugins' ),
 		name: 'plugin',
 		kind: 'root',
 		baseURL: '/wp/v2/plugins',
 		baseURLParams: { context: 'edit' },
 		key: 'plugin',
-	},
-];
+	} as const ),
+] as const;
+
+type Element = typeof defaultEntities[ number ];
+
+type DefaultEntityContext<
+	K extends string,
+	N extends string,
+	E = Extract< Element, { kind: K; name: N } >
+> = 'baseURLParams' extends keyof E
+	? 'context' extends keyof E[ 'baseURLParams' ]
+		? E[ 'baseURLParams' ][ 'context' ]
+		: 'view'
+	: 'view';
+
+type EntityKeyName<
+	K extends string,
+	N extends string,
+	E = Extract< Element, { kind: K; name: N } >
+> = 'key' extends keyof E ? E[ 'key' ] : 'id';
+
+type DeclaredEntity< K extends string, N extends string, RecordType > = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: EntityKeyName< K, N > extends keyof RecordType
+		? RecordType[ EntityKeyName< K, N > ]
+		: never;
+	defaultContext: DefaultEntityContext< K, N >;
+};
+type APIEntity<
+	K extends string,
+	N extends string,
+	RecordType,
+	KeyName = 'id',
+	C = 'edit'
+> = {
+	kind: K;
+	name: N;
+	recordType: RecordType;
+	keyType: KeyName extends keyof RecordType ? RecordType[ KeyName ] : never;
+	defaultContext: C;
+};
+
+export type EntityType< C extends EntityTypes.Context = any > =
+	| DeclaredEntity< 'root', 'site', EntityTypes.Settings< C > >
+	| DeclaredEntity< 'root', 'postType', EntityTypes.Type< C > >
+	| DeclaredEntity< 'root', 'media', EntityTypes.Attachment< C > >
+	| DeclaredEntity< 'root', 'taxonomy', EntityTypes.Taxonomy< C > >
+	| DeclaredEntity< 'root', 'sidebar', EntityTypes.Sidebar< C > >
+	| DeclaredEntity< 'root', 'widget', EntityTypes.Widget< C > >
+	| DeclaredEntity< 'root', 'widgetType', EntityTypes.WidgetType< C > >
+	| DeclaredEntity< 'root', 'user', EntityTypes.User< C > >
+	| DeclaredEntity< 'root', 'comment', EntityTypes.Comment< C > >
+	| DeclaredEntity< 'root', 'menu', EntityTypes.NavMenu< C > >
+	| DeclaredEntity< 'root', 'menuItem', EntityTypes.NavMenuItem< C > >
+	| DeclaredEntity< 'root', 'menuLocation', EntityTypes.MenuLocation< C > >
+	| DeclaredEntity<
+			'root',
+			'navigationArea',
+			EntityTypes.NavigationArea< C >
+	  >
+	| DeclaredEntity< 'root', 'theme', EntityTypes.Theme< C > >
+	| DeclaredEntity< 'root', 'plugin', EntityTypes.Plugin< C > >
+	| APIEntity< 'postType', 'post', EntityTypes.Post< C > >
+	| APIEntity< 'postType', 'page', EntityTypes.Page< C > >
+	| APIEntity< 'postType', 'wp_template', EntityTypes.WpTemplate< C > >
```

Comment:
are these extra kind and name fields inherent to the entity records?
if so, should we consider adding them in the entity record types directly so we don't have to duplicate the list here? or are these values specific to `getEntityRecord` and not particularly a part of the entity record itself?

Reply:
The API knows nothing about the kind and name fields, this is just a way for core-data to index them in Redux state.